### DISCUSSION TITLE
Switch to dark theme is not working on title bar

### DISCUSSION THREAD

**Discussion by jeff9571** [AvaloniaUI/Avalonia#12170] [Q&A] [answered]
here is the window axaml file:
`<Window xmlns="https://github.com/avaloniaui"
        xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml"
        xmlns:vm="using:v.ViewModels"
        xmlns:d="http://schemas.microsoft.com/expression/blend/2008"
        xmlns:mc="http://schemas.openxmlformats.org/markup-compatibility/2006"
        mc:Ignorable="d" d:DesignHeight="1080" d:DesignWidth="1920"
        x:Class="v.Views.MainWindow"
        Icon="/Assets/avalonia-logo.ico"
        Background="{DynamicResource BackgroundBrush1}"
        Title="Test" 
        WindowStartupLocation="CenterScreen"
        SizeChanged="Control_OnSizeChanged">`
only the window's content can be re-themed
![image](https://github.com/AvaloniaUI/Avalonia/assets/35749706/5c83e20c-0b53-417c-926d-92db06f73194)

**Comment by timunie** [comment, marked as answer]
In Windows 10 there is no (official) API to do it, so we cannot support it here. In Win11 it should work iirc. There is an undocumented Win23 hackfix, which you can do on your side. Search for it in WPF, should be applicable to Avalonia as well.

**Reply by jeff9571** [reply]
Yes, in Win 11, it works fine.

**Comment by steveberdy** [comment]
For anyone wanting a solution, here is my implementation (modified from https://stackoverflow.com/a/62811758)
It _has_ been documented on Microsoft docs after someone made a PR, so hopefully we'll get an official Avalonia implementation.

I implemented it in Avalonia specifically. It will set the title bar color to the theme color (light/dark) and will even change it if you change the theme. For some reason this only works for one change, after that you have to restart the app. I'm not sure what needs to be done to force a repaint of the window.

To implement for Avalonia:
```csharp
using System;
using System.Runtime.InteropServices;

public partial class TitleBar
{
    [LibraryImport("dwmapi.dll")]
    private static partial int DwmSetWindowAttribute(IntPtr hwnd, int attr, ref int attrValue, int attrSize);

    private const int DWMWA_USE_IMMERSIVE_DARK_MODE_BEFORE_20H1 = 19;
    private const int DWMWA_USE_IMMERSIVE_DARK_MODE = 20;
    private static readonly int darkModeAttr;

    static TitleBar()
    {
        darkModeAttr = IsWindows10OrGreater(18985)
            ? DWMWA_USE_IMMERSIVE_DARK_MODE
            : DWMWA_USE_IMMERSIVE_DARK_MODE_BEFORE_20H1;
    }

    public static bool UseImmersiveDarkMode(IntPtr handle, bool enabled)
    {
        if (IsWindows10OrGreater(17763))
        {
            int useImmersiveDarkMode = enabled ? 1 : 0;
            return DwmSetWindowAttribute(handle, darkModeAttr, ref useImmersiveDarkMode, sizeof(int)) == 0;
        }

        return false;
    }

    private static bool IsWindows10OrGreater(int build = -1)
    {
        return Environment.OSVersion.Version.Major >= 10 && Environment.OSVersion.Version.Build >= build;
    }
}
```

and then during initialization you'll write something like this

```csharp
if (window.TryGetPlatformHandle()?.Handle is { } h)
{
    TitleBar.UseImmersiveDarkMode(h, Application.Current.RequestedThemeVariant == Avalonia.Styling.ThemeVariant.Dark);
    window.ActualThemeVariantChanged += (_, _) =>
    {
        TitleBar.UseImmersiveDarkMode(h, Application.Current.RequestedThemeVariant == Avalonia.Styling.ThemeVariant.Dark);
        // do something to force repaint if necessary
    };
}
```
